### ISSUE TITLE
Realm database and integration

### ISSUE THREAD

**Issue by avaidyam** [ScienceSquad/Health#46]
Attach the Realm database to an Android Application context which is accessible to all modules, and ensure the modules' model layers are all in sync and properly annotated.

See [Realm's guide](https://realm.io/docs/java/latest/) on this.

_Note: Along with the `@Required` annotation, `@NonNull` should be used wherever possible!_

[Good reference on how to integrate along with RxJava.](https://github.com/kboyarshinov/realm-rxjava-example)
[Good reference from Realm IO using modules.](https://github.com/realm/realm-java/tree/master/examples/moduleExample)

TODO: 
- Set up a database for each Module such that they can access their via reads, writes, and queries. 
- Allow the database to build a calendar. 

---
- **Estimated Time:** _12 hours._
- **Prerequisites:** _None._

**Comment by avaidyam** [comment]
Suggested idea:
- `ModuleContext`
  - `applicationContext()`
    - provides Activity and Application backing.
  - `databaseContext()`
    - provides Realm backing.
  - `sensorContext()`
    - provides SensorManager backing.
  - `viewContext()`
    - available for View interaction if one is displayed.

**Comment by Brahmancer** [comment]
Could you explain more in detail what this would mean?

**Comment by avaidyam** [comment]
We went over this in detail in the last meeting, but we can go over it again in our next one, probably Sunday.

Essentially, a module would have an associated context binder; that's where all of the import and export of the model and view stuff would happen. This allows for extreme flexibility, implementation of MVC, and lots of nicer things we'll discuss later.

**Comment by avaidyam** [comment]
Here's a good example of what I mean.

``` java
public final class StepsModule extends Module {

    // when StepsModule starts, we need to observe if the user is running, cycling, etc.
    // we need to reflect their activity in our view, and possibly stop tracking steps temporarily
    public void start() {
        self.context.eventContext().subscribe(RunModule.ActivityChangedEvent.class, (event) -> {
            // we subscribe to events of type ActivityChangedEvent from RunModule
            // if RunModule ever alerts us that the activity type changed, this code will be called
            self.context.jobContext().queue(() -> {
                // we can queue an event to run on the UI thread when it's available to us from here
                // if we're on the UI thread, we know for sure viewContext is available to us
                self.context.viewContext().perform((StepsView view) -> {
                    view.color = Color.RED; // perhaps this means "run" or "walk" or something else
                });
            });
        });
    }

    // ... code here ...

    // something triggers a UI update where we update the number of steps walked the whole day
    public void updateUI() {
        self.context.viewContext().ifExists((StepsView view) -> {
            // we're being displayed on-screen if we have an associated view context
            List<StepsModel> today = self.context.dataContext().where(StepsModel.class).for(Date.today()); // pseudocode
            int totalStepsToday = self._aggregateSteps(today);
            view.message = "You walked " + totalStepsToday + " steps today!";
        }).orElse(() -> {
            // we're not on screen, but we can store this data or do something else with it later
        });
    }
}
```

**Comment by Brahmancer** [comment]
That sounds like a good idea. I understand the high level concept so we can discuss more of the nitty gritty details later.

**Comment by avaidyam** [comment]
Did my updated example make more sense? The decoupling is supposed to make modules reactive to changes on either end of MVC (as the Modules are the controllers themselves). Instead of directly messing with a view, you talk through its `viewContext()` which is a `binder` that is only available to you on the UI thread (we'll all learn how terrible threading is with UI stuff soon...) and sanitizes your requests to the view.

Now, suppose I want to rewrite my UI from scratch: I can. Without this adapter/binder layer, you can't. You're tightly coupled into the UI and you're quite frankly screwed. We'll go over this later on too. :shipit:

**Comment by Brahmancer** [comment]
It does from a backend stand point but I'm not really familiar with how it relates with the UI so it would probably best to go over in detail and in person how exactly the UI interacts with the back end :smile:

**Comment by avaidyam** [comment]
See, this is why you come to the meetings :wink: 

@colinrford @Wall-Dough we had a blast! :laughing:

**Comment by Brahmancer** [comment]
Again, I apologize for the short notice of me bailing on on the last meeting.

**Comment by avaidyam** [comment]
@Brahmancer lol that's fine! Make it to the next one :)

**Comment by Brahmancer** [comment]
Needed: 
An abstract DataBaseContext for database queries, reads, and writes
A RealmContext that extends it to work with Realm
An abstract RealmObject that allows for basis of individual Realm Databases for each Module?

**Comment by Brahmancer** [comment]
@avaidyam, this may seem like a silly question but here it goes. For a Context, this is something that would be implemented by us the developers? If so, is there a good guide on implementing just a general Context?

Edit: Brainwave. Would it be a bad idea to base the Realm Contexts off of the context at the start of the application?

**Comment by avaidyam** [comment]
@Brahmancer The `Context` as in the global [**Android** `Context` god-object](https://possiblemobile.com/2013/06/context/)? Or the definition of a "context" we'll be using? Literally, it means "the circumstances that form the setting for an event, statement, or idea, and in terms of which it can be fully understood and assessed."

**tl;dr** a `Context` is a "circumstance" object that, when available, allows events or computation to leave the "local" scope (the `Module`), _and_ the method by which the events or computation should be understood at the other end.

**tl;dr tl;dr** It's a thing that allows kids to _not only_ play inside a sandbox, but also invite other kids to play in their sandbox, or go to other sandboxes as well. In fact, if this kid went to a sandbox owned by a rich kid, versus one owned by a poor kid, the things he or she could play with would be vastly different -- and this is why we need to **ask** what our sandbox will be.

From that analogy, the **asking** would be called **dependency injection** (which we're not _really_ fully doing), and the `Module` would be the kid. Makes sense?

---

@Brahmancer Also, it's totally fine. If you speak of the Android `Context`, it's a god-object, which means quite literally, anything goes.

**Comment by Brahmancer** [comment]
I guess from how the application will run, we'll set up all the modules upon launching of the application right so that every module can grab that context. if need be.

**Comment by avaidyam** [comment]
@Brahmancer Better do it in a way that we never have to deal with that god-forsaken `Context` nonsense after the application startup.

**Comment by colinrford** [comment]
We can check this guy off the list! #55
